pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/generator/BaseTaskGenerator.java

```diff
@@ -131,4 +140,44 @@ public void generateTasks(List<TableConfig> tableConfigs, List<PinotTaskConfig>
   public String getMinionInstanceTag(TableConfig tableConfig) {
     return TaskGeneratorUtils.extractMinionInstanceTag(tableConfig, getTaskType());
   }
+
+  @Override
+  public boolean isAllowDownloadFromServer(TableConfig tableConfig) {
+    return TaskGeneratorUtils.extractMinionAllowDownloadFromServer(tableConfig, getTaskType());
+  }
+
+  public List<URI> getSegmentServerURIs(TableConfig tableConfig, String segmentName) {
+    String peerDownloadScheme = tableConfig.getValidationConfig().getPeerSegmentDownloadScheme();
+    List<URI> segmentServerURIs = PeerServerSegmentFinder.getPeerServerURIs(
+        _clusterInfoAccessor.getPinotHelixResourceManager().getHelixZkManager(),
+        tableConfig.getTableName(), segmentName, peerDownloadScheme);
+    Collections.shuffle(segmentServerURIs);
+    return segmentServerURIs;
+  }
+
+  public Map<String, String> getBaseTaskConfigs(TableConfig tableConfig, List<String> segmentNames) {
+    Map<String, String> baseConfigs = new HashMap<>();
+    baseConfigs.put(MinionConstants.TABLE_NAME_KEY, tableConfig.getTableName());
+    baseConfigs.put(MinionConstants.SEGMENT_NAME_KEY, StringUtils.join(segmentNames,
+          MinionConstants.SEGMENT_NAME_SEPARATOR));
+    Map<String, List<String>> segmentServerUriMap = new HashMap<>();
+    if (isAllowDownloadFromServer(tableConfig)) {
+      segmentServerUriMap = segmentNames.stream()
+          .collect(Collectors.toMap(
+              segmentName -> segmentName,
+              segmentName -> getSegmentServerURIs(tableConfig, segmentName)
```

Comment:
This is a good catch! 
Going into this if we overload this method `getPeerServerURIs`, we will have to update `getOnlineServersFromExternalView` to take in a list of segments and return the result accordingly. But seems we will lose the observability at segment level (most logs will help find issues at segment level).
As this would be an optimisation mainly for tasks where multiple segments are ran in one task (e.g. MergeRollupTask) do you think we should extend the scope here to update `PeerServerSegmentFinder` as well?

Reply:
Continuing the thought: does it make sense for us to precompute all segments' peers when generating the task? I think it should be done while downloading the segment, because EV can take time for segments to be updated. Could we do it in `downloadSegmentToLocal`?

Reply:
> Could we do it in downloadSegmentToLocal?

Yes looks like it might be possible there! Seems strange i was trying this before to do in executor but found some package dependency was missing. Now it seems doable suddenly. Let me update the PR. Thanks for the suggestion!

Reply:
Updated! Now the flow is simplified a lot!! We just check while executing the task itself if the config is set. If yes, we proceed to download from server else no! There are no major changes in task-generator flow anymore.